sklearn/neighbors/lof.py

```diff
@@ -0,0 +1,322 @@
+        WARNING: The argument X is supposed to contain new data. The samples in
+        X are not considered in any neighborhood.
+        The decision function on training data is available by considering the
+        opposite of the outlier_factor_ attribute.
+
+        Parameters
+        ----------
+
+        X : array-like, shape (n_samples, n_features), optional
+            The query sample or samples to compute the Local Outlier Factor
+            w.r.t. the training samples.
+
+        Returns
+        -------
+        lof_scores : array of shape (n_samples,)
+            The Local Outlier Factor of each input samples. The lower,
+            the more abnormal.
+        """
+        check_is_fitted(self, ["threshold_", "outlier_factor_",
+                               "_k_distance_value_fit_X_",
+                               "neighbors_indices_fit_X_"])
+
+        X = check_array(X, accept_sparse='csr')
+        return -self._local_outlier_factor(X)  # as bigger is better
+
+    def _k_distance(self, X=None):
+        """
+        Compute the k_distance and the neighborhood of query samples wrt
+        training samples self._fit_X.
+        If X=None, neighbors of each sample self._fit_X are returned.
+        In this case, the query point is not considered its own neighbor.
+        """
+        distances, neighbors_indices = self.kneighbors(
+            X=X, n_neighbors=self.n_neighbors)
+        k_dist = distances[:, self.n_neighbors - 1]
+
+        return k_dist, neighbors_indices
+
+    def _local_reachability_density(self, distances_X, neighbors_indices):
+        """The local reachability density (LRD)
+
+        The LRD of a sample is the inverse of the average reachability
+        distance of its k-nearest neighbors.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            The samples to compute the LRD w.r.t. self._fit_X
+            If None, compute the LRD of self._fit_X w.r.t. itself.
+            (In this case samples are not considered in their own
+            neighborhood)
+
+        neighbors_indices : array-like of shape (n_samples, self.n_neighbors)
+            neighbors indices of X (or of self._fit_X if X is None)
+            Added as a parameter to avoid re-computing it with
+            self._k_distance(X)[1] when X is not None (if X is None, it is
+            equal to self.neighbors_indices_fit_X_)
+
+        Returns
+        -------
+        local_reachability_density : array, shape (n_samples,)
+            The local reachability density of each sample.
+        """
+        dist_k = self._k_distance_value_fit_X_[neighbors_indices,
+                                               self.n_neighbors - 1]
+        reach_dist_array = np.maximum(distances_X, dist_k)
+
+        #  1e-10 to avoid `nan' when when nb of duplicates > n_neighbors:
+        return 1. / (np.mean(reach_dist_array, axis=1) + 1e-10)
+
+    def _local_outlier_factor(self, X=None):
```

Comment:
I think it will be easier to understand to give as parameter directly `distances_X` and `neighbors_indices_X`.
Like that we will make a real difference between calling the LOF with the inner X (in `fit`) or a new X (in `decision_function`).